File: src/services/revocation_service.py
from uuid import UUID
from typing import Optional
from datetime import datetime, timezone

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from src.config import Settings
from src.models.database.certificate import Certificate
from src.core.cache.redis_manager import RedisManager


class RevocationService:
    def __init__(self, redis_manager: RedisManager):
        self.redis = redis_manager

    async def revoke_certificate(
        self, certificate_id: UUID, reason: str, revoked_by: str, db: AsyncSession
    ) -> bool:
        """
        Revoke a certificate

        Returns:
            bool: True if revoked successfully, False if already revoked/expired
        """
        # Get certificate
        stmt = select(Certificate).where(Certificate.id == certificate_id)
        result = await db.execute(stmt)
        cert = result.scalar_one_or_none()

        if not cert:
            raise ValueError("Certificate not found")

        # Check if already revoked
        if cert.revoked_at is not None:
            return False

        # Check if already expired
        if cert.expires_at < datetime.now(timezone.utc):
            return False

        # Update certificate
        stmt = (
            update(Certificate)
            .where(Certificate.id == certificate_id)
            .values(revoked_at=datetime.now(timezone.utc), revocation_reason=reason)
        )

        await db.execute(stmt)
        await db.commit()

        # Invalidate cache
        await self.redis.invalidate_certificate(str(certificate_id))

        # Invalidate any active session tokens (optional)
        pattern = f"session:*:{certificate_id}"
        async for key in self.redis.client.scan_iter(pattern):
            await self.redis.client.delete(key)

        return True

    async def check_certificate_status(
        self, certificate_id: UUID, db: AsyncSession
    ) -> dict:
        """
        Check if certificate is valid, with caching
        """
        # Check cache first
        cached = await self.redis.get_certificate_status(str(certificate_id))

        if cached:
            return {
                "is_valid": cached["is_valid"],
                "expires_at": cached["expires_at"],
                "revoked_at": cached.get("revoked_at"),
                "source": "cache",
            }

        # Query database
        stmt = select(Certificate).where(Certificate.id == certificate_id)
        result = await db.execute(stmt)
        cert = result.scalar_one_or_none()

        if not cert:
            return {"is_valid": False, "error": "Certificate not found"}

        now = datetime.now(timezone.utc)
        is_valid = not cert.revoked_at and cert.expires_at > now

        # Cache result
        await self.redis.cache_certificate_status(
            str(certificate_id), is_valid, cert.expires_at
        )

        return {
            "is_valid": is_valid,
            "expires_at": cert.expires_at.isoformat(),
            "revoked_at": cert.revoked_at.isoformat() if cert.revoked_at else None,
            "revocation_reason": cert.revocation_reason,
            "source": "database",
        }

    async def bulk_revoke_by_issuer(
        self, issuer_id: UUID, reason: str, db: AsyncSession
    ) -> int:
        """
        Revoke all active certificates issued by a specific issuer

        Returns:
            int: Number of certificates revoked
        """
        now = datetime.now(timezone.utc)

        stmt = (
            update(Certificate)
            .where(
                Certificate.issuer_id == issuer_id,
                Certificate.revoked_at.is_(None),
                Certificate.expires_at > now,
            )
            .values(
                revoked_at=now, revocation_reason=f"Bulk revocation by issuer: {reason}"
            )
            .returning(Certificate.id)
        )

        result = await db.execute(stmt)
        await db.commit()

        revoked_ids = result.scalars().all()

        # Invalidate cache for all revoked certificates
        for cert_id in revoked_ids:
            await self.redis.invalidate_certificate(str(cert_id))

        return len(revoked_ids)

    async def reinstate_certificate(
        self, certificate_id: UUID, db: AsyncSession
    ) -> bool:
        """
        Reinstate a revoked certificate (admin only)
        """
        stmt = (
            update(Certificate)
            .where(
                Certificate.id == certificate_id, Certificate.revoked_at.is_not(None)
            )
            .values(revoked_at=None, revocation_reason=None)
        )

        result = await db.execute(stmt)
        await db.commit()

        if result.rowcount > 0:  # type: ignore
            await self.redis.invalidate_certificate(str(certificate_id))
            return True

        return False


def get_revocation_service():
    settings = Settings()
    redis_manager = RedisManager(settings.redis_url, settings.redis_cache_ttl_seconds)

    return RevocationService(redis_manager)

File: app/Services/RevocationService.php
<?php

namespace App\Services;

use App\Models\Certificate;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Redis;
use Carbon\Carbon;

class RevocationService
{
    protected int $cacheTtl;

    public function __construct()
    {
        // Fetch TTL from configurations; defaults to 1 hour if unassigned
        $this->cacheTtl = config('cache.certificate_ttl', 3600);
    }

    /**
     * Revoke a unique certificate instance.
     * 
     * @return bool True if revoked successfully, false if already revoked/expired
     */
    public function revokeCertificate(string $certificateId, string $reason, string $revokedBy): bool
    {
        $cert = Certificate::findOrFail($certificateId);

        // Check if already revoked
        if ($cert->revoked_at !== null) {
            return false;
        }

        // Check if already expired
        if ($cert->expires_at->isPast()) {
            return false;
        }

        return DB::transaction(function () use ($cert, $reason) {
            $cert->update([
                'revoked_at' => Carbon::now('UTC'),
                'revocation_reason' => $reason
            ]);

            // Invalidate application cache
            $this->invalidateCache($cert->id);

            // Purge any active server session keys tied to this certificate
            $this->purgeActiveSessions($cert->id);

            return true;
        });
    }

    /**
     * Check if a certificate is valid, utilizing highly responsive cache stores.
     */
    public function checkCertificateStatus(string $certificateId): array
    {
        $cacheKey = "certificate:status:{$certificateId}";

        // Attempt reading directly from Redis via the cache driver abstraction Layer
        $cached = Cache::get($cacheKey);

        if ($cached) {
            return [
                "is_valid" => $cached['is_valid'],
                "expires_at" => $cached['expires_at'],
                "revoked_at" => $cached['revoked_at'] ?? null,
                "source" => "cache"
            ];
        }

        $cert = Certificate::find($certificateId);

        if (!$cert) {
            return [
                "is_valid" => false,
                "error" => "Certificate not found"
            ];
        }

        $now = Carbon::now('UTC');
        $isValid = ($cert->revoked_at === null) && ($cert->expires_at->isAfter($now));

        $cacheData = [
            'is_valid' => $isValid,
            'expires_at' => $cert->expires_at->toIso8601String(),
            'revoked_at' => $cert->revoked_at?->toIso8601String()
        ];

        // Store evaluation state inside Redis
        Cache::put($cacheKey, $cacheData, $this->cacheTtl);

        return array_merge($cacheData, [
            "revocation_reason" => $cert->revocation_reason,
            "source" => "database"
        ]);
    }

    /**
     * Revoke all active credentials mapped to a singular Issuer entity.
     * 
     * @return int Total number of rows altered during atomic bulk operation
     */
    public function bulkRevokeByIssuer(string $issuerId, string $reason): int
    {
        $now = Carbon::now('UTC');

        return DB::transaction(function () use ($issuerId, $reason, $now) {
            // Find targets matching valid runtime parameters
            $targetQuery = Certificate::where('issuer_id', $issuerId)
                ->whereNull('revoked_at')
                ->where('expires_at', '>', $now);

            // Pluck matching targets into memory arrays to execute specific target flushes
            $revokedIds = $targetQuery->pluck('id')->toArray();

            if (empty($revokedIds)) {
                return 0;
            }

            // Perform rapid multi-row batch database update execution
            Certificate::whereIn('id', $revokedIds)->update([
                'revoked_at' => $now,
                'revocation_reason' => "Bulk revocation by issuer: {$reason}"
            ]);

            // Clean cache metrics across altered parameters
            foreach ($revokedIds as $id) {
                $this->invalidateCache($id);
            }

            return count($revokedIds);
        });
    }

    /**
     * Reinstates a closed or mistakenly revoked certificate profile back to operational use.
     */
    public function reinstateCertificate(string $certificateId): bool
    {
        return DB::transaction(function () use ($certificateId) {
            $affectedRows = Certificate::where('id', $certificateId)
                ->whereNotNull('revoked_at')
                ->update([
                    'revoked_at' => null,
                    'revocation_reason' => null
                ]);

            if ($affectedRows > 0) {
                $this->invalidateCache($certificateId);
                return true;
            }

            return false;
        });
    }

    /**
     * Centralized utility to handle multi-key structural cache resets.
     */
    protected function invalidateCache(string $certificateId): void
    {
        Cache::forget("certificate:status:{$certificateId}");
        Cache::forget("certificate:profile:{$certificateId}"); // Clear additional service tags if present
    }

    /**
     * Directly scan redis memory blocks to safely eliminate session traces.
     */
    protected function purgeActiveSessions(string $certificateId): void
    {
        $redisConnection = Redis::connection();
        $pattern = "session:*:[{$certificateId}]";

        // Utilize native non-blocking Redis SCAN cursor iteration profiles over keys loops
        $cursor = "0";
        do {
            [$cursor, $keys] = $redisConnection->scan($cursor, ['MATCH' => $pattern, 'COUNT' => 100]);
            
            if (!empty($keys)) {
                // Strip the application prefix if configured inside database setups before calling delete
                $cleanKeys = array_map(function ($key) {
                    return preg_replace('/^' . config('database.redis.options.prefix') . '/', '', $key);
                }, $keys);

                $redisConnection->del($cleanKeys);
            }
        } while ($cursor !== "0");
    }
}